src/likelihoods/negativebinomial.jl

```diff
@@ -1,36 +1,169 @@
+
+[^1] Hilbe, Joseph M. Negative binomial regression. Cambridge University Press, 2011.
+"""
+struct NegativeBinomialLikelihood{Tp<:NBParam,Tl<:AbstractLink} <: AbstractLikelihood
+    params::Tp # Likelihood parametrization (and parameters)
     invlink::Tl
+    function NegativeBinomialLikelihood(
+        params::Tparam, invlink=default_invlink(params)
+    ) where {Tparam<:NBParam}
+        # we convert `invlink` into a `Link` object if it's not the case already
+        invlink = link(invlink)
+        return new{Tparam,typeof(invlink)}(params, invlink)
+    end
 end
 
-function NegativeBinomialLikelihood(l=logistic; successes::Real=1)
-    return NegativeBinomialLikelihood(successes, link(l))
+@deprecate NegativeBinomialLikelihood(l=logistic; successes=1) NegativeBinomialLikelihood(
+    NBParamI(successes), l
+)
+
+function (l::NegativeBinomialLikelihood)(::Real)
+    return error(
+        "not implemented for type $(typeof(l)). For your custom type to run you ",
+        "need to implement `(l::NegativeBinomialLikelihood{<:MyNBParam})(f::Real)`. ",
+        "For a full explanation, see `NegativeBinomialLikelihood` docs",
+    )
 end
 
 @functor NegativeBinomialLikelihood
 
-(l::NegativeBinomialLikelihood)(f::Real) = NegativeBinomial(l.successes, l.invlink(f))
-
 (l::NegativeBinomialLikelihood)(fs::AbstractVector{<:Real}) = Product(map(l, fs))
+
+@doc raw"""
+    NBParamSuccess(successes)
+
+Negative Binomial parametrization with `successes` the number of successes and
+`invlink(f)` the probability of success.
+This corresponds to the definition used by [Distributions.jl](https://juliastats.org/Distributions.jl/latest/univariate/#Distributions.NegativeBinomial).
+
+```math
+  p(y|\text{successes}, p=invlink(f)) = \frac{\Gamma(y+\text{successes})}{y! \Gamma(\text{successes})} p^\text{successes} (1 - p)^y
+```
+"""
+struct NBParamSuccess{T} <: NBParamProb
+    successes::T
+end
+
+function (l::NegativeBinomialLikelihood{<:NBParamSuccess})(f::Real)
+    return NegativeBinomial(l.params.successes, l.invlink(f))
+end
+
+@doc raw"""
+    NBParamFailure(failures)
+
+Negative Binomial parametrization with `failures` the number of failures and
+`invlink(f)` the probability of success.
+This corresponds to the definition used by [Wikipedia](https://en.wikipedia.org/wiki/Negative_binomial_distribution).
+
+```math
+  p(y|\text{failures}, p=\text{invlink}(f)) = \frac{\Gamma(y+\text{failures})}{y! \Gamma(\text{failures})} p^y (1 - p)^{\text{failures}}
+```
+"""
+struct NBParamFailure{T} <: NBParamProb
+    failures::T
+end
+
+function (l::NegativeBinomialLikelihood{<:NBParamFailure})(f::Real)
+    return NegativeBinomial(l.params.failures, 1 - l.invlink(f))
+end
+
+# Helper function to convert mean and variance to p and r
+_nb_mean_excessvar_to_r_p(μ::Real, ev::Real) = μ / ev, 1 / (1 + ev)
+
+"""
+    NBParamI(α)
+
+Negative Binomial parametrization with mean `μ=invlink(f)` and variance `v=μ(1 + α)`.
```

Comment:
nit: I think this looks marginally neater / more readable in the rendered docs:
```suggestion
Negative Binomial parametrization with mean `μ = invlink(f)` and variance `v = μ(1 + α)`.
```